### PR TITLE
IGNITE-22519: Sql. Numerics. Conversion produces incorrect results

### PR DESCRIPTION
**Pull request by lowka**
Converts numeric literals into correct values w/ overflow checks. 

https://issues.apache.org/jira/browse/IGNITE-22519

BIGINT case (a cast from decimal string to BIGINT) should be resolved under - https://issues.apache.org/jira/browse/IGNITE-20889

Enabled tests muted under https://issues.apache.org/jira/browse/IGNITE-22519.

---
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/integrationTest/sql/aggregate/group/test_group_by_distinct.test [hunk 1]
```diff
@@ -29,20 +29,20 @@ SELECT val FROM integers GROUP BY ALL GROUPING SETS ((val), (val))
 4
 
 query R rowsort
-SELECT val FROM ( VALUES (1), (1.0::REAL), (1.0::REAL), (2::DOUBLE), (1.1::REAL), (2::DECIMAL(2)) ) t(val) GROUP BY ALL GROUPING SETS ((val), (val))
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/sql/aggregate/group/test_group_by_distinct.test:32]: `1.1` does not have the exact representation as a double (we get `1.100000023841858`) , use `1.5` instead

File: modules/sql-engine/src/integrationTest/sql/set/test_intersect.test [hunk 2]
```diff
@@ -74,7 +74,7 @@ statement ok
 CREATE TABLE t_float(f1 float);
 
 statement ok
-INSERT INTO t_float(f1) VALUES (-1.2345678901234e+200), (-34.84),  (123), (456), (456.1), (-456);
+INSERT INTO t_float(f1) VALUES (-1.2345678901234e+10), (-34.84),  (123), (456), (456.1), (-456);
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/sql/set/test_intersect.test:77]: This value can not be represented as a float value but constant literals didn't overflow prior to this  patch.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java [hunk 3]
```diff
@@ -366,7 +366,7 @@ public void testDecimalLiteral() {
                 .returns(new BigDecimal("1.00")).check();
 
         assertQuery("SELECT CASE WHEN false THEN DECIMAL '1.00' ELSE DECIMAL '0.0' END")
-                .returns(new BigDecimal("0.0")).check();
+                .returns(new BigDecimal("0.00")).check();
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java:369]: Type system reports DECIMAL(3,2), so `0.00` is correct

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/RexToLixTranslator.java [hunk 4]
```diff
@@ -912,7 +913,7 @@ public static Expression translateLiteral(
                 final Primitive primitive = Primitive.ofBoxOr(javaClass);
                 final Comparable value = literal.getValueAs(Comparable.class);
                 if (primitive != null && value instanceof Number) {
-                    value2 = primitive.number((Number) value);
+                    value2 = Primitives.convertPrimitiveExact(primitive, (Number) value);
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/RexToLixTranslator.java:916]: Primitive::number does not perform overchecks for Float/Double .

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 5]
```diff
@@ -114,14 +114,14 @@ public class RexUtils {
...
-    private static final BigDecimal MIN_DOUBLE_VALUE = BigDecimal.valueOf(Double.MIN_VALUE);
-
     private static final BigDecimal MAX_DOUBLE_VALUE = BigDecimal.valueOf(Double.MAX_VALUE);
 
-    private static final BigDecimal MIN_FLOAT_VALUE = BigDecimal.valueOf(Float.MIN_VALUE);
+    private static final BigDecimal MIN_DOUBLE_VALUE = MAX_DOUBLE_VALUE.negate();
 
     private static final BigDecimal MAX_FLOAT_VALUE = BigDecimal.valueOf(Float.MAX_VALUE);
 
+    private static final BigDecimal MIN_FLOAT_VALUE = MAX_FLOAT_VALUE.negate();
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:123]: Float.MIN_VALUE/Double.MIN_VALUE are not minimum values for those types.

**Review by zstan — APPROVED** [review]

**Review by korlov42 — APPROVED** [review]

**Pull request merged**